### PR TITLE
Create dev setup script

### LINKED ISSUE #153: Read local ~/.kube/config if the CLUSTER_* creds are missing
This will help with local development, and make it easier to get started.

First thought on approaching this - in Config.py, it reads the CLUSTER_* environment variables. After the apply() method has been called in __init__(), if the CLUSTER_* variables are null, attempt to read the local Kubernetes config file. Usage of pyyaml is expected (to parse and read).

Another thought would be to move over to the Kubernetes async Python client altogether, since this is pretty much duplicate code from that library.

### PR DESCRIPTION
**Pull request by rewop**
This PR creates a `asyncy-server dev_setup` script that creates needed resources in kubernetes, and creates a .env file with environment variables set.

What the script doesn't do yet:
- it doesn't initialize postgresql as for the database repo
- it doesn't accept arguments yet

resolve #153

### CODE REVIEW

File: asyncy/Service.py [hunk 1]
```diff
@@ -88,6 +90,71 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
         logger.log_raw('info', 'Shutdown complete!')
 
     @staticmethod
+    @main.command()
+    def dev_setup():
+        """Sets up development environment to run platform engine.
+        It creates a .env file with the value to be ued to run the engine.
+        This command assumes that kubectl is installed and connected to
+        a cluster and that postgres is installed and reachable through
+        localhost:5432
```
Comment (judepereira) [inline asyncy/Service.py]: @rewop - Can you accept the host and port for this as a parameter?

File: asyncy/Service.py [hunk 2]
```diff
@@ -88,6 +90,71 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
         logger.log_raw('info', 'Shutdown complete!')
 
     @staticmethod
+    @main.command()
+    def dev_setup():
+        """Sets up development environment to run platform engine.
+        It creates a .env file with the value to be ued to run the engine.
+        This command assumes that kubectl is installed and connected to
```
Comment (judepereira) [inline asyncy/Service.py]: Add a test for kubectl being installed and available.

File: asyncy/Service.py [hunk 3]
```diff
@@ -88,6 +90,71 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+        a cluster and that postgres is installed and reachable through
+        localhost:5432
+        """
+
+        # load configuration from default
+        kube_config.load_kube_config()
+        kube_config_inst = kube_config.kube_config.Configuration()
+
+        # Create resources needed in the cluster
+        # @todo these resources will be created in the namespace default
```
Comment (judepereira) [inline asyncy/Service.py]: Make use of the namespace `asyncy-system`.

File: asyncy/Service.py [hunk 4]
```diff
@@ -88,6 +90,71 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+        # Should we also create a namespace for the engine?
+        kube_resources = [
+            'service_accounts/engine.yaml',
+            'secrets/engine.yaml',
+            'role_bindings/engine.yaml',
+        ]
+        command = ('curl -s '
```
Comment (judepereira) [inline asyncy/Service.py]: Is it possible to use python `requests` to fetch the file, and then apply it with the `kubernetes` python library?

File: asyncy/Service.py [hunk 5]
```diff
@@ -88,6 +90,71 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+                   'master/kubernetes-pre-init/{resource}'
+                   ' | kubectl apply -f -')
+        for resource in kube_resources:
+            click.echo(f'creating or updating resource {resource}')
+            resp = subprocess.run(command.format(
+                resource=resource), shell=True)
+            resp.check_returncode()
+
+        # CLUSTER_HOST
+        host = kube_config_inst.host
```
Comment (judepereira) [inline asyncy/Service.py]: Need to show a picker here before selecting the cluster to be used. Incase the cluster selected is some production cluster, we do not want to accidentally mess with it.

We need to make this CLI interactive, and ask for a Y/n explicitly.

File: asyncy/Service.py [hunk 6]
```diff
@@ -88,6 +90,71 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+        # POSTGRES
+        # @todo we could parametrize these or adjust the
+        # defaults in config.py to use 'asyncy' db instead of
+        # 'postgres'
+        postgres_options = ('options=--search_path='
+                            'app_public,app_hidden,app_private,public '
+                            'dbname=asyncy '
+                            'user=postgres')
+
+        env_file = open('.env', 'w')
```
Comment (judepereira) [inline asyncy/Service.py]: Can you add the corresponding code to accept a path to an `env` file?

File: asyncy/Service.py (hunk 4 above)
Comment (rewop) [reply asyncy/Service.py]: I tried, but I couldn't find any counterpart of `kubectl apply` command from the api. The `apply` command doesn't perform any action if the resources are not updated.

I could of course check if resources are already created, but I thought this was easier and cleaner. What do you think?

**Comment by rewop** [comment]
@judepereira I addressed your comments. Could you check it again? Thanks

File: asyncy/Service.py (hunk 4 above)
Comment (judepereira) [reply asyncy/Service.py]: Hmm, okay. In that case, we can just rely on `kubectl`.

File: asyncy/Service.py [hunk 7]
```diff
@@ -88,6 +89,96 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+    @main.command()
+    @click.option('--psql-host',
+                  help='Hostname to connect to postgresql',
+                  default='localhost')
+    @click.option('--psql-port',
+                  help='Hostname to connect to postgresql',
+                  default='5432')
+    @click.argument('env_file', type=click.File('w'), default='.env')
+    def dev_setup(psql_host, psql_port, env_file):
+        """Sets up development environment to run platform engine.
```
Comment (judepereira) [inline asyncy/Service.py]:
```suggestion
        """
        Sets up development environment to run platform engine.
```

File: asyncy/Service.py [hunk 8]
```diff
@@ -88,6 +89,96 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+        It creates a .env file with the value to be ued to run the engine.
```
Comment (judepereira) [inline asyncy/Service.py]:
```suggestion
        It creates a .env file with the value to be used to run the engine.
```

File: asyncy/Service.py [hunk 9]
```diff
@@ -88,6 +89,96 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+        """
+
+        from shutil import which
+
+        # Make sure kubectl is installed
+        if which('kubectl') is None:
+            print(('Missing kubectl cli. Follow '
```
Comment (judepereira) [inline asyncy/Service.py]: Use click.echo().

File: asyncy/Service.py [hunk 10]
```diff
@@ -88,6 +89,96 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+                   'to install it first.'), file=sys.stderr)
+            exit(1)
+
+        # Load configuration from default
+        kube_config.load_kube_config()
+        kube_config_inst = kube_config.kube_config.Configuration()
+
+        # CLUSTER_HOST
+        host = kube_config_inst.host
+        click.echo(f'ATTENTION: resources will be create in cluster at {host}')
```
Comment (judepereira) [inline asyncy/Service.py]:
```suggestion
        click.echo(f'ATTENTION: resources will be created in cluster at {host}')
```

File: asyncy/Service.py [hunk 11]
```diff
@@ -88,6 +89,96 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+                   'master/kubernetes-pre-init/{resource}'
+                   ' | kubectl apply --namespace asyncy-system -f -')
+        for resource in kube_resources:
+            click.echo(f'creating or updating resource {resource}')
+            resp = subprocess.run(command.format(
+                resource=resource), shell=True)
+            resp.check_returncode()
+
+        # CLUSTER_CERT
+        cert_file = open(kube_config_inst.cert_file, "r")
```
Comment (judepereira) [inline asyncy/Service.py]:
```suggestion
        cert_file = open(kube_config_inst.cert_file, 'r')
```

File: asyncy/Service.py [hunk 12]
```diff
@@ -88,6 +89,96 @@ def start(port, debug, sentry_dsn, release, prometheus_port):
...
+        cert = cert_file.read().replace("\n", "\\n")
```
Comment (judepereira) [inline asyncy/Service.py]:
```suggestion
        cert = cert_file.read().replace('\n', '\\n')
```

File: asyncy/Service.py (hunk 4 above)
Comment (rewop) [reply asyncy/Service.py]: What I could do maybe is using the same approach I use for namespace. Create the resource and check for conflict error. Though i would have to request the yaml files and parse then, which would make the code a bit more verbose.

**Comment by judepereira** [comment]
@rewop Nice work! 

What do you think about checking to see if the database is reachable and bootstrapped? And if not, print some message to clone asyncy/database, run the bootstrap. Thoughts on doing this?

This script looks almost complete.

**Some todos:**
- [ ] Move the dev command to another file alongside Service.py, say CliDevEnvHelper.py
- [ ] Write unit tests
- [ ] Ensure Circle CI tests succeed

File: asyncy/Service.py (hunk 4 above)
Comment (judepereira) [reply asyncy/Service.py]: Naa, don't worry about it. It's much cleaner this way.

File: asyncy/Service.py (hunk 9 above)
Comment (rewop) [reply asyncy/Service.py]: I use `print` to print on stderr. I haven't found a way to do so with click

**Comment by rewop** [comment]
@judepereira thanks to check again.

> What do you think about checking to see if the database is reachable and bootstrapped? And if not, print some message to clone asyncy/database, run the bootstrap. Thoughts on doing this?

That could be done of course. I just have 2 thoughts:
1. what check would prove if the db is bootstrapped? 
2. I assume a developer when using this script is following a getting started before using this command, which should give the db bootstrapped. The goal of the script is to create the `.env` file in the end, and the db is not an hard requirement. Shouldn't a developer be skilled enough to have the db correctly set? 

About writing unit tests, I am not sure of the value of it. The command is just a sequence of I/O operations. I would be mainly mocking all those I/O and check they have been called, but this is done by just running the script. I am not sure the effort is worth the value. What do you think?

I will move the code in a different command.

**Comment by rewop** [comment]
@judepereira can you give a look again? I fixed the errors from flake in CI though now there is a new error that looks like some network problem, and I cannot restart it.

I will wait your answer about my comment on test before proceeding with them.

File: asyncy/Service.py (hunk 9 above)
Comment (judepereira) [reply asyncy/Service.py]: `click.echo('Hello World!', err=True)`

**Comment by judepereira** [comment]
> what check would prove if the db is bootstrapped?

The table `releases` exists.

> I assume a developer when using this script is following a getting started before using this 
command, which should give the db bootstrapped. The goal of the script is to create the .env file in the end, and the db is not an hard requirement. Shouldn't a developer be skilled enough to have the db correctly set?

I agree, however, we must strive to make this bullet proof. Somebody who wants to just play around and somehow reaches here directly, shouldn't be frustrated (obviously, the error will make it obvious that the connection to the DB failed, but we should detect and prevent it if possible). Also note - the `.env` file contains PostgreSQL creds, hence we should validate that the engine can connect to it, and that it's bootstrapped.

> I fixed the errors from flake in CI though now there is a new error that looks like some network problem, and I cannot restart it

Yes, it did seem to be a one off issue. I just triggered it again, and it worked.

> About writing unit tests, I am not sure of the value of it. The command is just a sequence of I/O operations. I would be mainly mocking all those I/O and check they have been called, but this is done by just running the script. I am not sure the effort is worth the value. What do you think?

As of now, we cannot say with confidence that the script will remain the same throughout the lifetime of this project. Somebody is bound to contribute to it in the future, enhance it, or even change it significantly. Hence, to ensure that what was working continues to work when somebody else contributes, IMO asserting that the call sequence + what's being written to the `.env`
 is indeed the right output produced from the right command.

**Comment by rewop** [comment]
@judepereira thanks to take time to respond. I will create tests. I will let you know when it is done.

**Comment by judepereira** [comment]
@rewop Are tests the only thing pending here? Can I test drive this locally?

**Comment by rewop** [comment]
Yes tests are the only part missing.

**Comment by judepereira** [comment]
Closing as this PR is now stale.

**Pull request closed without merging**